pkg/api/validation/validation.go

```diff
@@ -2480,9 +2551,10 @@ func ValidatePodSpec(spec *api.PodSpec, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 
 	allVolumes, vErrs := ValidateVolumes(spec.Volumes, fldPath.Child("volumes"))
+	allDeviceVolumes := GetClaimVolumes(spec.Volumes)
```

Comment:
I would prefer validating the volume type in ValidateVolumeDevices itself instead of reducing the list of volumes.  If you reduce the list of volumes, users could see an error message like "volume not found", which is confusing.

Reply:
volume not found would only happen if a user supplies a volumeDevice in the container spec that doesn't have a corresponding volume.name + source of PersistentVolumeClaim which is expected behavior.  Maybe I'm missing something?  Or are you saying that you don't like the way the current validation works for all other validation by passing in lists (names, paths, etc...)?

Reply:
Consider this scenario:
```
volumeDevices:
   - name: vol1
     devicePath: /some/path
volumes:
  - name: vol1
    hostpath:
      path: /some/device
  - name: vol2
    persistentVolumeClaim:
      claimName: some-other-block-PVC
```

The error message the user is going to get is "vol1 not found in volumes list", instead of a message like "Only PVCs are allowed for block devices"

Reply:
hmmm...based on that scenario and what they would get based on the pushed code is the result of this check:  0 volumes matching name + pvc source and devices do exist

```
	if len(volumes) == 0 && devices != nil {
		allErrs = append(allErrs, field.Required(fldPath.Child("volumeDevices"), "Pod spec volumes must contain a valid PersistentVolumeClaim to use with volumeDevices"))
		return allErrs
	}
```
wording different than yours but seems to be accomplishing same thing (probably need to update the wording a bit to add a few more details for users).  I was just following the current code pattern by using lists.

Reply:
I see your point, I'll need to pass in the volume spec directly and cross-reference to get to that level of granularity.  thanks